### ISSUE TITLE
Nominatim methods throw NoMethodError when @data['address'] is nil

### ISSUE THREAD

**Issue by duarme** [alexreisner/geocoder#1556]
### Expected behavior

When calling `result.city`, if there is no `city`, [Nominatim#city](https://github.com/alexreisner/geocoder/blob/master/lib/geocoder/results/nominatim.rb#L26) should return nil.

### Actual behavior

When `@data['address']` is nil, it throws a **NoMethodError: undefined method `key?' for nil:NilClass**:

![CleanShot 2021-12-27 at 11 30 41](https://user-images.githubusercontent.com/794154/147467731-f78935a9-c41f-4d56-99ce-ca2de3449ea2.png)

### Steps to reproduce

Try to geocode any result that doesn't have `@data['address']` and call `Nominatim#city` on it (or any other similar method).

### Environment info

* Geocoder version: 1.6.7
* Rails version:  5.2.4.3
* Database (if applicable): postgres
* Lookup (if applicable): locationiq

### Proposed solution

How about switching 

```ruby
[...] if @data['address'].key?(key)
```
 
 to 
 
 ```ruby
 [...] if @data['address']&.key?(key)
 ``` 
 
 into [nominatim.rb](https://github.com/alexreisner/geocoder/blob/master/lib/geocoder/results/nominatim.rb) ?

**Comment by alexreisner** [comment]
Thanks for this. Your proposed solution will work for the `city` method but there are more than a dozen others that rely on `@data['address']` being non-nil, and this would be a good time to fix all of them. Out of curiosity, do you have a sample Nominatim response where geocoding is successful but address is empty? Would be good to add a [test fixture](https://github.com/alexreisner/geocoder/tree/master/test/fixtures).

**Comment by duarme** [comment]
You're welcome. Oddly enough, I don't, I couldn't replicate this because every time I retry, it works. It seems a glitch in the lookup service, in order to catch it I should add some debug line and run them in production. Anyway, I'm not sure that's what's going on. I mean, how do you know that geocoding is successful? Before the failing line I'm calling `result.latitude` and `result.longitude`, but they could be nil and I'd get `nil.to_f => 0.0`, won't I?

```ruby
      def coordinates
        [@data['latitude'].to_f, @data['longitude'].to_f]
      end

      def latitude
        coordinates[0]
      end

      def longitude
        coordinates[1]
      end
```

Let me find a Search with `latitude = 0.0`:

```
irb(main):019:0> s = Search.find(147274455)
  Search Load (1.2ms)  SELECT  "searches".* FROM "searches" WHERE "searches"."id" = $1 LIMIT $2  [["id", 147274455], ["LIMIT", 1]]
=> #<Search id: 147274455, query: "editing+video", geo_query: "falconara+marittima+(marche)", geocoded_query: nil, latitude: 0.0, longitude: 0.0, [...], region: nil, geocoded_region: nil, geocoded_city: nil>
irb(main):020:0> s.geocoded?
=> true # <= woah, this is bad 👀 
irb(main):021:0> s.geocode
Traceback (most recent call last):
        2: from (irb):21
        1: from app/models/search.rb:71:in `block in <class:Search>'
NoMethodError (undefined method `key?' for nil:NilClass)
irb(main):022:0> 
```

Yep, and what's worse `s.geocoded?` returns `true`. Let's see the result:

```
irb(main):023:0> results = Geocoder.search(s.geo_query)
=> [#<Geocoder::Result::LocationIq:0x00007fb03b5ca330 @data={"error"=>"Unable to geocode"}, @cache_hit=false>]
irb(main):024:0> results.first
=> #<Geocoder::Result::LocationIq:0x00007fb03b5ca330 @data={"error"=>"Unable to geocode"}, @cache_hit=false>
```

Okay, now it seems something that I should solve on my part too. I should check whether there is any error before saving the returned latitude, longitude, city, etc. 

Still, IMHO I think that the geocoder gem behavior can be improved. Specifically, `object.geocoded?` probably shouldn't return true after this error, and `result.city` and the other dozen methods shouldn't throw that error, they should return `nil` or raise a proper exception instead. 

What's your take on this? 🙂

**Comment by xtr3me** [comment]
I made a PR with some changes that should improve the lookup of addresses with Nominatim. What is your take on this @alexreisner?

**Comment by alexreisner** [comment]
This should be resolved by #1562. @duarme let me know if you still have issues after the fix is applied.
